csrc/fusion_segmenter.cpp

```diff
@@ -316,16 +437,149 @@ std::unique_ptr<SegmentedFusion> SegmentedFusion::fromCompleteFusion(
+  FUSER_PERF_SCOPE("SegmentedFusion::deserialize");
+  NVF_ERROR(buffer != nullptr, "serde::SegmentedFusion is nullptr.");
+
+  // NOTE SchedulerEntry::proposeHeuristics can add values and expressions to
+  // the fusion. We relax the constraints here because we already know the
+  // proposed scheduler for each segmented group.
+  const std::deque<Val*>& vals = complete_fusion_->deterministic_vals();
+  const std::deque<Expr*>& exprs = complete_fusion_->deterministic_exprs();
+  NVF_ERROR(
+      complete_fusion_->vals().size() <= buffer->num_vals(),
+      "The complete fusion has ",
+      vals.size(),
+      " values while serialization expected ",
+      buffer->num_vals(),
+      " values.");
+  NVF_ERROR(
+      complete_fusion_->exprs().size() <= buffer->num_exprs(),
+      "The complete fusion has ",
+      exprs.size(),
+      " expressions while serialization expected ",
+      buffer->num_exprs(),
+      " expressions.");
+
+  segmented_fusion_name_ = buffer->segmented_fusion_name();
+
+  // Construct segmented groups first because they are necessary for the
+  // segmented edge's constructor
+  // NOTE: Use regular for-loop to avoid unused variable ‘idx’ error
+  for (size_t idx = 0; idx < buffer->groups()->size(); ++idx) {
+    newGroup();
+  }
+
+  // Create segmented edges
+  for (auto idx : c10::irange(buffer->edges()->size())) {
+    auto se_fb = buffer->edges()->Get(idx);
+    newEdge(
+        groups_.at(se_fb->from_segmented_group()),
+        groups_.at(se_fb->to_segmented_group()),
+        vals.at(se_fb->val()));
+  }
+
+  // Deserialize segmented groups
+  for (auto idx : c10::irange(buffer->groups()->size())) {
+    auto sg_fb = buffer->groups()->Get(idx);
+    groups_.at(idx)->deserialize(sg_fb, vals, exprs, groups_, edges_);
+  }
+
+  for (auto idx : *buffer->force_fp16_tv_set()) {
+    auto val = vals.at(idx);
+    NVF_CHECK(
+        val->isA<TensorView>(),
+        "Segmented Fusion Deserialization: Expected Val to be a TensorView.");
+    force_fp16_tv_set_.emplace(val->as<TensorView>());
+  }
+
+  force_half_precision_type_ =
+      DataType(static_cast<PrimDataType>(buffer->force_half_precision_type()));
+
+  finalize();
+}
+
+flatbuffers::Offset<serde::SegmentedEdge> SegmentedFusion::serialize(
+    flatbuffers::FlatBufferBuilder& builder,
+    const nvfuser::SegmentedEdge* edge,
+    const std::unordered_map<Val*, int64_t>& vals_map,
+    const std::unordered_map<SegmentedGroup*, int64_t>& groups_map) const {
+  FUSER_PERF_SCOPE("SegmentedEdge::serialize");
+  return serde::CreateSegmentedEdge(
+      builder,
+      groups_map.at(edge->from),
+      groups_map.at(edge->to),
+      vals_map.at(edge->val));
+}
+
+nvfuser::SegmentedEdge SegmentedFusion::deserialize(
+    const serde::SegmentedEdge* buffer,
+    const std::deque<Val*>& vals) {
+  FUSER_PERF_SCOPE("SegmentedEdge::deserialize");
+  NVF_ERROR(buffer != nullptr, "serde::SegmentedEdge is nullptr.");
+  return {
```

Comment:
I'd `NVF_ERROR` that `groups_` is populated before calling this function, so the contract of order is clear.